### PR TITLE
Add missing net_buf_unref for non-recoverable errors while sending l2cap data

### PR DESCRIPTION
**Pull request by prathje**
In case the connection breaks while transmitting, the net_buf while be dequeued but not unreferenced.

### CODE REVIEW

**Review by jhedberg — CHANGES_REQUESTED** [review]
Looks fine, but you’ll need to have a proper prefix to the commit message summary (CI will probably also fail because of it). Something like “Bluetooth: L2CAP: Fix missing net_buf_unref()” should be sufficient, with more details in the commit message body.

**Comment by prathje** [comment]
I updated the commit message and also removed the trailing whitespace to hopefully get through compliance.

**Comment by prathje** [comment]
Sign-off was missing for compliance.

**Review by jhedberg — DISMISSED** [review]

**Review by Thalley — APPROVED** [review]

**Comment by jhedberg** [comment]
Does `bt_l2cap_chan_send()` require a similar fix (end of the function)?

**Comment by jhedberg** [comment]
@Vudentz @joerchan would be good if you guys could take a look, since this brings back memories of the ref-counting discussion we had a few months ago 😄

**Comment by Thalley** [comment]
> @Vudentz @joerchan would be good if you guys could take a look, since this brings back memories of the ref-counting discussion we had a few months ago

@jhedberg Perhaps we should add the DNM label if we want their input first

**Comment by jhedberg** [comment]
> @jhedberg Perhaps we should add the DNM label if we want their input first

@Thalley I dismissed my own review until then, which does the same trick in practice (can't merge now)

**Comment by joerchan** [comment]
> Vudentz joerchan would be good if you guys could take a look, since this brings back memories of the ref-counting discussion we had a few months ago

So this is a regression due to that change. I suggest to add this comment above the net_buf_unref .
```
				/* The ownership of the buffer is with the stack
				 * since bt_l2cap_chan_send has returned success
				 */
```

**Comment by prathje** [comment]
@joerchan I think a comment won't hurt, so I just went ahead and added it 👍

**Review by joerchan — APPROVED** [review]

**Comment by joerchan** [comment]
@prathje Thanks. I think you could squash the two commits though.

File: subsys/bluetooth/host/l2cap.c
Comment (jhedberg) [inline subsys/bluetooth/host/l2cap.c:879]: Sorry, but this is a bit confusing to me. On two levels.

Firstly, these branches handle the return value of l2cap_chan_le_send_sdu(), so why the reference to bt_l2cap_chan_send() in the comment? Was that a typo or intentional?

Secondly, "since bt_l2cap_chan_send has returned success" looks very odd considering that we're in a branch where l2cap_chan_le_send_sdu() has *not* returned success. I'd have worded it the other way around "If l2cap_chan_le_send_sdu() fails the ownership of the buffer is retained by the caller, so we must unref it" or something like that.

Comment (prathje) [reply subsys/bluetooth/host/l2cap.c:879]: Yes, this is intentional. The ownership has been transferred by a previous user call of `bt_l2cap_chan_send` (not the internal `l2cap_chan_le_send_sdu`). `bt_l2cap_chan_send` returned success to the caller, otherwise the buffer would not have been queued.

**Comment by prathje** [comment]
> @prathje Thanks. I think you could squash the two commits though.

Yes, I have thought of that but decided not to as the first commit has been merged already (https://github.com/nrfconnect/sdk-zephyr/pull/555). I am happy to merge if you still prefer it?

**Comment by carlescufi** [comment]
> > @prathje Thanks. I think you could squash the two commits though.
> 
> Yes, I have thought of that but decided not to as the first commit has been merged already ([nrfconnect/sdk-zephyr#555](https://github.com/nrfconnect/sdk-zephyr/pull/555)). I am happy to merge if you still prefer it?

That has not been merged to Zephyr, this commit was cherry-picked into a fork of the Zephyr repo.

**Comment by joerchan** [comment]
@prathje That's not a problem, what Carles said is correct, so please squash them. But thank you for being considerate.
If you keep the first commit message it will go smoothly.

File: subsys/bluetooth/host/l2cap.c
Comment (carlescufi) [inline subsys/bluetooth/host/l2cap.c:877]: Please remove the comment, since this is self-explanatory by the fact that this is an internal processing function that takes net_buf references from an internal queue, and therefore those are owned by the stack and not the application.

**Review by carlescufi — DISMISSED** [review]

File: subsys/bluetooth/host/l2cap.c
Comment (joerchan) [reply subsys/bluetooth/host/l2cap.c:877]: Ok, so the comment generated more confusion than clarity, so we discussed it and decided it was best to just remove it.

Comment (prathje) [reply subsys/bluetooth/host/l2cap.c:877]: Will do :)

**Comment by prathje** [comment]
@carlescufi @joerchan Thanks! With the requested comment removal, we are back to just the first commit anyway. I also did a quick rebase 👍

File: subsys/bluetooth/host/l2cap.c
Comment (prathje) [reply subsys/bluetooth/host/l2cap.c]: Comment was removed again. So I resolve this :)

**Review by jhedberg — APPROVED** [review]

**Pull request merged**
